tests/form_params_test.go

```diff
@@ -0,0 +1,385 @@
+		// to fail intentionally to pinpoint behaviour we regard as inconsistent
+		t.Skip()
+		t.Run(test.name, func(t *testing.T) {
+			resp, err := requesttesting.MakeRequest(context.Background(), test.req, func(req *http.Request) {
+				if contentLen := req.Header["Content-Length"][0]; contentLen != test.want.contentLen {
+					t.Errorf(`req.Header["Content-Length"]: got %v, want %v`, contentLen, test.want.contentLen)
+				}
+				if val := req.FormValue("veggie"); val != test.want.queryParamVal {
+					t.Errorf("req.FormValue: want potato, got %s", val)
+				}
+			})
+			if err != nil {
+				t.Fatalf("MakeRequest(): got err %v, want nil", err)
+			}
+			if !!bytes.HasPrefix(resp, []byte(statusOKReq)) {
+				t.Errorf("response status: got %s, want %s", resp, statusOKReq)
+			}
+		})
+	}
+
+	postReq := []byte("POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: application/x-www-form-urlencoded; charset=ASCII\r\n" +
+		"Content-Length: 13\r\n" +
+		"Content-Length: 12\r\n" +
+		"\r\n" +
+		"veggie=potato\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if contentLen := req.Header["Content-Length"][0]; contentLen != "13" {
+			t.Error("MakeRequest(): request should have been rejected when list of values is given to Content-Length, rather inconsistent if you ask me.")
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(statusBadReq)) {
+		t.Errorf("response status: got %s, want %s", resp, statusBadReq)
+	}
+}
+
+// Test whether form parameters with Content-Type:
+// application/x-www-form-urlencoded that break percent-encoding will be
+// ignored
+func TestFormParametersBreakUrlEncoding(t *testing.T) {
+	postReq := []byte("POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: application/x-www-form-urlencoded\r\n" +
+		"Content-Length: 10\r\n" +
+		"\r\n" +
+		"veggie=%sc\r\n" +
+		"\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), postReq, func(req *http.Request) {
+		if val := req.Form["veggie"]; len(req.Form["veggie"]) != 0 {
+			t.Errorf("req.Form[\"veggie\"]: got %s, want %s.", val, "")
+		}
+	})
+	if err != nil {
+		t.Fatalf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !bytes.HasPrefix(resp, []byte(statusOKReq)) {
+		t.Errorf("response status: got %s, want %s", resp, statusOKReq)
+	}
+}
+
+func TestBasicMultipartForm(t *testing.T) {
+	reqBody := "--123\r\n" +
+		"Content-Disposition: form-data; name=\"foo\"\r\n" +
+		"\r\n" +
+		"bar\r\n" +
+		"--123--\r\n"
+	postReq := "POST / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Type: multipart/form-data; boundary=\"123\"\r\n" +
+		"Content-Length: " + strconv.Itoa(len(reqBody)) + "\r\n" +
+		"\r\n" +
+		reqBody + "\r\n" +
+		"\r\n"
+	resp, err := requesttesting.MakeRequest(context.Background(), []byte(postReq), func(req *http.Request) {
+		e := req.ParseMultipartForm(1024)
```

Comment:
nit: use "err" for error variables. Also, you can do: `if err := req.ParseMultipartForm(1024); err != nil {` in one line.